### PR TITLE
Reinterpret between 32 and 16 bit texture formats

### LINKED ISSUE #9018: Rendering issue in Spongebob Yellow Avenger
once in game the 2.5D sections render weirdly but the 3D sections (inside houses) work perfectly
ules00280_00000

### LINKED ISSUE #9604: MX vs. ATV Reflex (USA) ULUS10429 graphics problem
v1.4-80-gd2172c7
it looks like  it has the same problem as #7459 
happens with both d3d11 and ogl 
ulus10429_00001

### LINKED ISSUE #9850: White overlap in Cars Race-o-Rama
A white rectangle overlaps almost all the screen during a race. Had no luck in fixing it by playing with graphics settings. Some game objects appear over this overlap but track and enemies are invisible. System: Windows 8.1, Asus X540S.with Pentium N3700, Intel HD Graphics.

Race screenshot

### LINKED ISSUE #15898: Rendering issues in Tantalus Media games (Spongebob, MX ATV, etc)
As Gamemulatorer pointed out in https://github.com/hrydgard/ppsspp/issues/9018#issuecomment-911914556 , these games have weird rendering techniques in common.

All except MegaMind (and it seems MX vs ATV: On The Edge) render a very complicated striped post-processing effect, that doesn't actually seems to do very much (EDIT: it applies textured depth fog).

The details seem to vary a bit from game to game but all have in common copying the screen in small stripes to a temporary and doing strange processing there, involving depth texturing and reinterpret between 16 and 32-bit formats. I created this common issue where I'll write up any progress and link PRs to.

We are currently very unsuccessful emulating this, but I think it's not quite as hard as it looks.

Features I think we need to implement for this to work: 

- [x] #15892
- [x] Detect use of a framebuffer living in the "stride gap" of another one and adjust its width accordingly (and make sure it doesn't get detected as the same texture as the big one)
- [x] Probably smarter interactions between block copies and framebuffers, taking the new sequence numbers into account etc. In the end, we might to share framebuffer match detection logic between block copies, textures and framebuffers...
- [x] 32-bit<->16-bit format reinterpret: #15907

Still will likely not perform very well, but better than nothing.

Details are in the various sub-issues:

- #9850
- #9604
- #9018

### PR DESCRIPTION
**Pull request by hrydgard**
Fixes #15898 (Rendering issues in Tantalus games), more details in that issue.

Fixes #9850
Fixes #9604
Fixes #9018

Fixes the God of War shadow regression reported in comments to #15888.

Hopefully this also gets us closer to universally enabling reinterpret.

Not too afraid of merging this despite its size, can't find any regressions but if there are, they should be easy to fix, easier than before due to many things being handled in ways that are easier to reason about.

Spongebob now renders near perfectly, the other games have minor issues at 2x+ resolution, and apparently some subtle off by 1. Good enough for a first merge.

~~So far, Spongebob is rendering better but still broken due to two issues:~~

- [x] Need to implement the depth texture swizzle of +200000 addresses, since the game expects it
  - [x] However, the current implementation somehow breaks Me and My Katamari's depth fog
    * It seems that the color format affects the 200000-type swizzle. It may not be present when the color format of the framebuffer is 16-bit. This will need hardware testing (does make sense though to have different swizzles for different color depth, to have the depth access pattern as different as possible from color).
- [x] ~Something is wrong with the depth texture depal, seems the shift is wrong, or somehow the input to it, something, as it causes an annoying repeating gradient~ EDIT: Turns out that the game uses a stencil/alpha trick to clear alpha, making this not apply where it's out of range.
  CLUT mode is set to CLUT16, mask 0xFF, shift 2. Depth data is 16-bit, so that just takes bits 2-10. The high bits 11-16 are unused in the lookup, causing a repeated gradient, which is weird and looks bad.
  

Other things to do:

- [ ] Cars / MX vs ATV games have some different issues
  - [x] Framebuffers cloned for reinterpret created too small
  - [x] Half the screen still corrupt
  - [x] Minor display corruption all over (UV precision issue?)
  - [ ] Thin stripes across the screen (maybe higher res is not possible? but something else is 1px off..)
- [ ] GE dumps of Spongebob don't render correctly, might be interesting to figure out why and fix
- [x] Double check the first two commits

image
image

### CODE REVIEW

**Comment by hrydgard** [comment]
I've tested this on mobile now, and it works just fine on Adreno, except that other than Spongebob, the games render a bit dark, for whatever reason. I'm gonna leave that to later fixes.

File: GPU/D3D11/StateMappingD3D11.cpp [hunk 1]
```diff
@@ -293,6 +293,34 @@ void DrawEngineD3D11::ApplyDrawState(int prim) {
...
 				dynState_.useStencil = true;
 				dynState_.stencilRef = stencilState.testRef;
+
+				// Nasty special case for Spongebob and similar where it tries to write zeros to alpha/stencil during
+				// depth-fail. We can't write to alpha then because the pixel is killed. However, we can invert the depth
+				// test and modify the alpha function...
+				if (gstate.isDepthTestEnabled() && !gstate.isDepthWriteEnabled() && stencilState.zFail == GE_STENCILOP_ZERO &&
+					stencilState.sFail == GE_STENCILOP_KEEP && stencilState.zPass == GE_STENCILOP_KEEP &&
+					stencilState.testFunc == GE_COMP_ALWAYS &&
+					stencilState.writeMask == 0xFF && stencilState.testMask == 0xFF && stencilState.testRef == 0xFF) {
```
Comment (unknownbrackets) [inline GPU/D3D11/StateMappingD3D11.cpp:303]: testMask/testRef ought to be irrelevant if it's ALWAYS.  But don't we need to verify the actual depth test itself?  What about if it's writing color?  How can we safely inverse the depth test without caring about those?

-[Unknown]

**Review by unknownbrackets — COMMENTED** [review]
I feel like I'm missing something that makes the depth swap safe - what if a game were using that configuration of parameters but actually intended to draw color?  Maybe left from a previous draw.

Also feel like we should probably just add those cmds to the depth dirty, or else we'll find little bugs...

-[Unknown]

File: GPU/Common/FramebufferManagerCommon.cpp [hunk 2]
```diff
@@ -680,13 +680,13 @@ void FramebufferManagerCommon::CopyToColorFromOverlappingFramebuffers(VirtualFra
...
+				sources.push_back(CopySource{ src, RASTER_COLOR, 0, 0 });
 			} else if (dst->fb_stride == src->fb_stride * 2) {
 				// Reinterpret from 32-bit to 16-bit.
-				WARN_LOG_N_TIMES(i32to16, 50, G3D, "16-bit to 32-bit reinterpret needed: %s to %s", GeBufferFormatToString(src->fb_format), GeBufferFormatToString(dst->fb_format));
+				sources.push_back(CopySource{ src, RASTER_COLOR, 0, 0 });
 			} else {
-				// 16-to-16 reinterpret, surely.
-				WARN_LOG_N_TIMES(i16to16, 50, G3D, "16-bit reinterpret needed: %s to %s", GeBufferFormatToString(src->fb_format), GeBufferFormatToString(dst->fb_format));
+				// 16-to-16 reinterpret, should have been caught above already.
+				WARN_LOG(G3D, "Reinterpret: Shouldn't get here");
```
[the remark this answers is not in the document]
Comment (hrydgard) [reply GPU/Common/FramebufferManagerCommon.cpp:672]: Dunno, maybe should even assert, as it's a logical error.

File: GPU/D3D11/StateMappingD3D11.cpp (hunk 1 above)
Comment (hrydgard) [reply GPU/D3D11/StateMappingD3D11.cpp:303]: You're right, I got way too excited about it working and forgot to finish the checks. Will do.

File: GPU/Common/FramebufferManagerCommon.cpp [hunk 3]
```diff
@@ -1608,68 +1609,118 @@ bool FramebufferManagerCommon::NotifyFramebufferCopy(u32 src, u32 dst, int size,
...
+
+	// We work entirely in bytes when we do the matching, because games don't consistently use bpps that match
+	// that of their buffers. Then after matching we try to map the copy to the simplest operation that does
+	// what we need.
 
 	for (auto vfb : vfbs_) {
 		const u32 vfb_address = vfb->fb_address & 0x3FFFFFFF;
 		const u32 vfb_size = ColorBufferByteSize(vfb);
+
+		if (basePtr < vfb_address || basePtr > vfb_address + vfb_size) {
```
[the remark this answers is not in the document]
Comment (hrydgard) [reply GPU/Common/FramebufferManagerCommon.cpp:1654]: Oops, will fix.

File: GPU/Common/FramebufferManagerCommon.cpp [hunk 4]
```diff
@@ -1608,68 +1609,118 @@ bool FramebufferManagerCommon::NotifyFramebufferCopy(u32 src, u32 dst, int size,
...
+		// Some games use mismatching bitdepths. But make sure the stride matches.
+		// If it doesn't, generally this means we detected the framebuffer with too large a height.
+		// Use bufferHeight in case of buffers that resize up and down often per frame (Valkyrie Profile.)
+
+		// If it's outside the vfb by a single pixel, we currently disregard it.
+		if (memYOffset > vfb->bufferHeight - h) {
+			continue;
+		}
+
+		if (byteOffset == vfb->WidthInBytes() && vfb->WidthInBytes() < vfb->FbStrideInBytes()) {
```
[the remark this answers is not in the document]
Comment (hrydgard) [reply GPU/Common/FramebufferManagerCommon.cpp:1679]: Oops, checking the wrong thing here.... Fixing

File: GPU/D3D11/StateMappingD3D11.cpp (hunk 1 above)
Comment (hrydgard) [reply GPU/D3D11/StateMappingD3D11.cpp]: testMask is irrelevant yes but testRef gets written - although you're right, with KEEP/KEEP/ZERO nothing is reading it.

I'll remove the checks for those too.

**Comment by hrydgard** [comment]
As for adding the blend stuff to depth dirty, yes, technically it's needed. Still I'm reluctant because that is common enough state that it might have some impact, and setting stencil to KEEP/KEEP/ZERO like this is quite unusual so unlikely to be leftover state from some previous draw... But I might end up eating those words, heh.

File: GPU/Common/FramebufferManagerCommon.cpp [hunk 5]
```diff
@@ -664,12 +664,12 @@ void FramebufferManagerCommon::CopyToColorFromOverlappingFramebuffers(VirtualFra
...
 				// Reinterpret from 16-bit to 32-bit.
 				sources.push_back(CopySource{ src, RASTER_COLOR, 0, 0 });
-			} else if (dst->fb_stride == src->fb_stride * 2) {
+			} else if (src->fb_stride * 2 == dst->fb_stride) {
 				// Reinterpret from 32-bit to 16-bit.
 				sources.push_back(CopySource{ src, RASTER_COLOR, 0, 0 });
 			} else {
 				// 16-to-16 reinterpret, should have been caught above already.
-				WARN_LOG(G3D, "Reinterpret: Shouldn't get here");
+				_assert_("Reinterpret: Shouldn't get here");
```
Comment (unknownbrackets) [inline GPU/Common/FramebufferManagerCommon.cpp:672]:
```suggestion
				_assert_msg_(false, "Reinterpret: Shouldn't get here");
```

-[Unknown]

File: GPU/Common/GPUStateUtils.h [hunk 6]
```diff
@@ -207,7 +207,6 @@ inline bool SpongebobDepthInverseConditions(const GenericStencilFuncState &stenc
 		gstate.getDepthTestFunction() == GE_COMP_GEQUAL &&
 		stencilState.zFail == GE_STENCILOP_ZERO && stencilState.sFail == GE_STENCILOP_KEEP && stencilState.zPass == GE_STENCILOP_KEEP &&
 		stencilState.testFunc == GE_COMP_ALWAYS &&
-		stencilState.writeMask == 0xFF && stencilState.testMask == 0xFF && stencilState.testRef == 0xFF &&
```
Comment (unknownbrackets) [inline GPU/Common/GPUStateUtils.h]: `writeMask` should matter, even for 0?

Also, `gstate.getMaterialAmbientA() == 0x0` depends on the material update flag right?  It seems like just checking these values might still be unsafe, although this is probably a pretty narrow case at this point.

-[Unknown]

**Review by unknownbrackets — COMMENTED** [review]
State tracking bugs are always annoying, though you're right it should be pretty rare with these checks.

-[Unknown]

File: GPU/Common/FramebufferManagerCommon.cpp (hunk 5 above)
Comment (hrydgard) [reply GPU/Common/FramebufferManagerCommon.cpp:672]: haha doh

**Pull request merged**